### DISCUSSION TITLE
Global Buffer which is used for ReadBytes?

### DISCUSSION THREAD

**Discussion by SnazzyBlanket** [pschatzmann/arduino-audio-tools#1543] [Q&A] [answered]
``I am working on a project where I am using audio tools with my own stream classes. However I would prefer to use whatever global variable calls buffer in ReadBytes (in some code it would be called DMA_BUFFER but I am not sure what it is in audio Tools

i would rather not have to initi buffer within readBytes and if there is a global variable that already calls this it would save memory.

thanks!!

here is the current code

size_t readBytes(uint8_t* buffer, size_t length) override {
    int16_t* output = (int16_t*) buffer;
    for (size_t i = 0; i < length / sizeof(int16_t); i += 2) {
        float leftSig = 0, rightSig = 0;
        for (int j = 0; j < MAX_VOICES; j++) {
            if (!voices[j].env.isActive()) {  // Use the envelope's isActive method
                ReInitVoice(j);  // Re-trigger inactive voices
            }
            float outSig = 0;
            voices[j].Process(outSig);
            leftSig += outSig;  // Accumulate for left channel
            rightSig += outSig; // Accumulate for right channel (modify as needed)
        }
        leftSig = std::min(std::max(leftSig, -1.0f), 1.0f);  // Clamp the signal to prevent overflow
        rightSig = std::min(std::max(rightSig, -1.0f), 1.0f); // Clamp the signal to prevent overflow
        output[i] = static_cast<int16_t>(leftSig * 32767);  // Convert float to int16_t for audio output (left channel)
        output[i + 1] = static_cast<int16_t>(rightSig * 32767);  // Convert float to int16_t for audio output (right channel)
    }
    return length;
}

**Comment by pschatzmann** [comment, marked as answer]
I don't quite understand your question and your example code does not provide any context information about the overall data flow: It  pretty much depends on what you are using and your platform. In any case there is no global memory in the AudioTools and it is up to each class if some additional buffering is needed.
E.g. If you use a I2SStream of an ESP32 the reading comes from DMA memory and when you write the data goes to DMA memory, (if you use 2 channels) but this is an implementation detail of the API provided by the ESP32 and not of my functionality. 
 
I just see that you are doing a conversion from float: You could save some complexity by using an EncodedAudioStream with a 
DecoderFloat or use the integration that is provided for Maximilian.
